src/link-preview-expander.ts

```diff
@@ -1,16 +1,142 @@
 import * as sourcegraph from 'sourcegraph'
+import { checkIsURL, cleanURL, getWord } from './util'
+import parse from 'node-html-parser'
 
-export function activate(ctx: sourcegraph.ExtensionContext): void {
-    ctx.subscriptions.add(
+export function activate(context: sourcegraph.ExtensionContext): void {
+    context.subscriptions.add(
         sourcegraph.languages.registerHoverProvider(['*'], {
-            provideHover: () => ({
-                contents: {
-                    value: 'Hello world from link-preview-expander! 🎉🎉🎉',
-                    kind: sourcegraph.MarkupKind.Markdown
+            provideHover: (document, position) => {
+                const range = document.getWordRangeAtPosition(position)
+                if (!range) {
+                    return null
                 }
-            }),
+                const maybeURL = cleanURL(getWord(document, range))
+                const isURL = checkIsURL(maybeURL)
+
+                if (!isURL) {
+                    return null
+                }
+
+                /**
+                 * Creates the markdown string to be rendered in the hover tooltip.
+                 */
+                const createResult: (metadata?: Metadata) => sourcegraph.Badged<sourcegraph.Hover> = metadata => {
+                    const { image, title, description } = metadata ?? {}
+
+                    let markdownContent = `#### [${title || maybeURL}](${maybeURL})\n\n`
+
+                    if (image || description) {
+                        markdownContent += '---\n\n'
+
+                        if (image) {
+                            markdownContent += `<img height="64" src="${image}" align="left" />`
+                        }
+
+                        if (description) {
+                            markdownContent += `<p>${description ?? ''}</p>`
+                        }
+                    }
+
+                    return {
+                        contents: {
+                            value: markdownContent,
+                            kind: sourcegraph.MarkupKind.Markdown,
+                        },
+                    }
+                }
+
+                // TODO(tj): Return async iterable (once allowed) instead of promise so that we can show link before metadata loads
+                return fetch('https://cors-anywhere.herokuapp.com/' + maybeURL, { cache: 'force-cache' })
+                    .then(response => response.text())
```

Comment:
I think we should be throwing an error here if `!response.ok`. Otherwise, if it's a link to a private repo for example, we might show a link preview for the 401 sign in page, or the 404 page, and even cache it.

Reply:
Any reason to not use `async`/`await`? I think that is easier to understand for people less familiar with JS

Reply:
Yeah, it should throw an error, but should be caught in this extension to display a plain link. Good catch regarding `async` / `await`